src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,179 @@
+#include "software/ai/passing/pass_generator.h"
+
+#include <iomanip>
+
+#include "software/geom/algorithms/contains.h"
+#include "software/logger/logger.h"
+
+PassGenerator::PassGenerator(const TbotsProto::PassingConfig& passing_config)
+    : optimizer_(optimizer_param_weights),
+      random_num_gen_(RNG_SEED),
+      passing_config_(passing_config)
+{
+}
+
+PassWithRating PassGenerator::getBestPass(const World& world,
+                                          const std::vector<RobotId>& robots_to_ignore)
+{
+    auto receiving_positions_map =
+        sampleReceivingPositionsPerRobot(world, robots_to_ignore);
+
+    // if there are no friendly robots, return early
+    if (receiving_positions_map.empty())
+    {
+        // default pass with 0 rating
+        return PassWithRating{Pass(Point(), Point(), 1.0), 0};
+    }
+
+    // Optimize the receiving positions for each robot and get the best pass
+    PassWithRating best_pass = optimizeReceivingPositions(world, receiving_positions_map);
+
+    // Visualize the sampled passes and the best pass
+    if (passing_config_.pass_gen_vis_config().visualize_sampled_passes())
+    {
+        std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+        for (const auto& [robot_id, receiving_positions] : receiving_positions_map)
+        {
+            for (const Point& receiving_position : receiving_positions)
+            {
+                debug_shapes.push_back(*createDebugShape(
+                    Stadium(world.friendlyTeam().getRobotById(robot_id)->position(),
+                            receiving_position, 0.02),
+                    std::to_string(debug_shapes.size()) + "pg"));
+            }
+        }
+        std::stringstream stream;
+        stream << "BP:" << std::fixed << std::setprecision(3) << best_pass.rating;
+        debug_shapes.push_back(
+            *createDebugShape(Circle(best_pass.pass.receiverPoint(), 0.05),
+                              std::to_string(debug_shapes.size()) + "pg", stream.str()));
+        LOG(VISUALIZE) << *createDebugShapes(debug_shapes);
+    }
+
+    // Generate sample passes across the field for cost visualization
+    if (passing_config_.cost_vis_config().generate_sample_passes())
+    {
+        samplePassesForVisualization(world, passing_config_, best_pass.pass);
+    }
+
+    return best_pass;
+}
+
+std::map<RobotId, std::vector<Point>> PassGenerator::sampleReceivingPositionsPerRobot(
```

Comment:
You can completely ignore this comment: 

Actually, I thought about this a bit more. I don't think `std::map<RobotId, std::Vector<Point>>` is a good data structure for something like this. I was thinking more along the line of `std::vector<Point>`.  This is because `std::vector` is a lot more cache friendly since the memory is contiguous.

I remember there was a conversation being had where passing takes up a lot of time. Packing data tightly would allow for faster iteration. See: https://gamedev.stackexchange.com/questions/200076/i-dont-get-why-ecs-is-considered-more-performant-than-oop 

We could do something like the following; 

```
[ Robot_Id_One (N elements) .... Robot_Id_Two_Points (N elements) ... Robot_Id_Three_Points (N elements)]
```

Of course, it is a bit more complicated than that since we have ignored some robots. 

Anyway, you can complete ignore this comment because this may be premature optimization which is the root of all evil .

Reply:
Also, shouldn't this return a type of `std::map<robotid, std::vector<point>>&&` to use movement semantics? We could avoid a lot of unnecessary copying? 

I am not quite sure, but I think all the `std::vector<Point>`in the map  would be copied to line 29?

Reply:
nvm this is dangerous

Reply:
I agree with you that we could return an array, with a static number of elements being associated with the points associated with each robot and that would likely help with caching, but in terms of code readability, that would make the code much harder to understand. 

In terms of performance, this would be a rather minor gain as well, since the pass generator spends most of its time in the `ratePass` call, and the copying of return values generally doesn't impact performance that much. You can look at a profile of our code using KCacheGrind and observe that the copy constructor of this data structure would likely be called very minimally. 

In fact, C++ compilers have gotten fairly smart about not copying the return type in scenarios like this. 
https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization

I actually tried moving the return value using `std::move` and got the following error:
```
software/ai/passing/pass_generator.cpp:118:21: error: moving a local object in a return statement prevents copy elision [-Werror=pessimizing-move]
  118 |     return std::move(receiving_positions_map);
      |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
```